### ISSUE TITLE
Does this wrapper use ANSI, UTF8, or UTF16 or something else?

### ISSUE THREAD

**Issue by bwmarrin** [JuliaDB/ODBC.jl#168]
TL;DR - Does this package send query strings untouched or convert to UTF8/UTF16 or something else?

The long story..

So, I just started with Julia 3 days ago and am barely getting my toes wet :)  I've installed it, installed this, and got a connection working to a IBM AS/400 mainframe!  At the same time I've also been working to get the same connection working with Google Go and wasn't having any luck.  Since this package and the Go package are both essentially wrappers around the unixODBC C API I thought I might be able to find the difference making one work and the other not.

With the Go odbc package I was getting translation problems.  You can see all the details here that show the errors, debugging we've done, and environment details.. https://github.com/alexbrainman/odbc/issues/88

I tried to dig around in the ODBC code for this package to see what it does differently and I'm just too new to the language to follow it completely.  I saw references to `transcode` and `utf16` but I'm not sure exactly if those are used for everything or special cases.. So, I'm curious what's being done differently here and if this package just hands the SQL strings over without modification, or converts to UTF8/UTF16 or anything else?  Or maybe it's some entirely different difference..  /shrug

Either way, thanks for making this package! I think I'll definitely keep playing with Julia to test out some database work.

**Comment by quinnj** [comment]
hey @bwmarrin, welcome to Julia! Hopefully things work out well for you; while still young and fast evolving, I find it immensely satisfying to have fast code so accessible.

Here's how we handle it in ODBC:

* When we execute the query, the database driver will report whether they're returning UTF8, UTF16, or even UTF32 (this is also somewhat OS & ODBC driver dependent, i.e. there are differences between windows/OSX/Linux and unixODBC/iODBC/odbc32 (windows))
* When fetching the query, we return the data as-is, i.e. as UTF8, UTF16, or UTF32
* Once the data is fetched, we're utilizing the DataStreams.jl package to allow the results to be "output" into any format you want, CSV.jl, SQLite.jl, Feather.jl, DataFrame.jl, etc.
* When streaming the data to whatever output, we then do the `transcode` function to convert the raw string data in UTF8, UTF16, or UTF32 into the standard Julia String which is UTF8; so if the data is already UTF8, it's a no-op and we just put the data in a String, otherwise the raw data is transcoded first then wrapped in a String.

Hopefully that helps!

**Comment by bwmarrin** [comment]
Thank you for the quick response and all the details.  Really helps me understand better!

So, I think the Go package it is sending UTF16 for everything.  With your driver, you're asking first "What is supported" then using the appropriate encoding.  I bet that's the key difference.  Can you point me to the section of your code that asks the driver what encoding is supported? Maybe I can understand it well enough to make a PR for the Go package :)

**Comment by quinnj** [comment]
Hey @bwmarrin, sorry for the slow response! I've been travelling to JuliaCon in California this week, so I'm back online now.

So first we detect based on OS & ODBC driver [here](https://github.com/JuliaDB/ODBC.jl/blob/master/src/types.jl#L39), I then call `SQLDescribeCol` [here](https://github.com/JuliaDB/ODBC.jl/blob/master/src/Source.jl#L124) to get all the column information, which then uses the responses to get the final Julia types that are stored [here](https://github.com/JuliaDB/ODBC.jl/blob/master/src/types.jl#L414). Hopefully that points you in the right direction?

**Comment by bwmarrin** [comment]
Thanks for the help! I've looked at these links and am trying to understand them better.  The part I'm mostly curious about is where you said..

> When we execute the query, the database driver will report whether they're returning UTF8, UTF16, or even UTF32

Where is it that it reports the supported UTF encoding?

**Comment by quinnj** [comment]
So I slightly misspoke, the database drivers will report
```julia
const SQL_CHAR          = Int16(  1) # Character string of fixed string length n.
const SQL_VARCHAR       = Int16( 12) # Variable-length character string with a maximum string length n.
const SQL_LONGVARCHAR   = Int16( -1) # Variable length character data. Maximum length is data source–dependent.
const SQL_WCHAR         = Int16( -8) # Unicode character string of fixed string length n
const SQL_WVARCHAR      = Int16( -9) # Unicode variable-length character string with a maximum string length n
const SQL_WLONGVARCHAR  = Int16(-10) # Unicode variable-length character data. Maximum length is data source–dependent
```
one of the above, which map to the other code I mentioned to you which is ODBC driver manager/OS-specific. (the code above starts at line 179 in `src/types.jl`)
